File: 12-password-robusta/php/goodpassword.php
<?php

function good_password($password) {
    $length = strlen($password);
    if ($length < 8) {
        return false;
    }
    $has_upper = false;
    $has_lower = false;
    $has_digit = false;
    for ($i = 0; $i < $length; $i++) {
        $char = $password[$i];
        if (ctype_upper($char)) {
            $has_upper = true;
        } elseif (ctype_lower($char)) {
            $has_lower = true;
        } elseif (ctype_digit($char)) {
            $has_digit = true;
        }
    }
    return $has_upper && $has_lower && $has_digit;
}

$password = readline("Introduzca una contraseña: ");
if (good_password($password)) {
    echo "Contraseña válida\n";
} else {
    echo "La contraseña debe tener al menos 8 caracteres, una mayúscula, una minúscula, un número y un carácter especial\n";
}

File: 12-password-robusta/python/goodpassword.py
def good_password(password):
    if len(password) < 8:
        return False
    if not any(c.isupper() for c in password):
        return False
    if not any(c.islower() for c in password):
        return False
    if not any(c.isdigit() for c in password):
        return False
    if not any(c in "!#$%&/()=?¿¡*+-.:;,_-|@€" for c in password):
        return False
    return True


password = input("Introduzca una contraseña: ")
if good_password(password):
    print("Contraseña válida")
else:
    print("La contraeña debe tener al menos 8 caracteres, una mayúscula, una minúscula, un número y un carácter especial")